examples/console/agent.go

```diff
@@ -0,0 +1,370 @@
+	if err != nil {
+		return fmt.Errorf("proposing payment %d: %w", amountValue, err)
+	}
+	enc := json.NewEncoder(io.MultiWriter(a.conn, a.LogWriter))
+	err = enc.Encode(message{Update: &ca})
+	if err != nil {
+		return fmt.Errorf("sending payment: %w", err)
+	}
+	return nil
+}
+
+func (a *Agent) StartClose() error {
+	if a.conn == nil {
+		return fmt.Errorf("not connected")
+	}
+	if a.channel == nil {
+		return fmt.Errorf("not introduced")
+	}
+	// Submit declaration tx
+	declTx, closeTx, err := a.channel.CloseTxs()
+	if err != nil {
+		return fmt.Errorf("building declaration tx: %w", err)
+	}
+	err = a.Submitter.SubmitFeeBumpTx(declTx)
+	if err != nil {
+		return fmt.Errorf("submitting declaration tx: %w", err)
+	}
+	// Revising agreement to close early
+	ca, err := a.channel.ProposeClose()
+	if err != nil {
+		return fmt.Errorf("proposing the close: %w", err)
+	}
+	timeStart := time.Now()
+	closeReady := make(chan bool)
+	go func() {
+		enc := json.NewEncoder(io.MultiWriter(a.Conn(), a.LogWriter))
+		dec := json.NewDecoder(io.TeeReader(a.Conn(), a.LogWriter))
+		err = enc.Encode(message{Close: &ca})
+		if err != nil {
+			fmt.Fprintf(a.LogWriter, "error: sending the close proposal: %v\n", err)
+		}
+		err = a.conn.SetReadDeadline(time.Now().Add(observationPeriodTime))
+		if err != nil {
+			fmt.Fprintf(a.LogWriter, "error: setting read deadline of conn: %v\n", err)
+		}
+		authorized := false
+		m := message{}
+		err = dec.Decode(&m)
+		if errors.Is(err, os.ErrDeadlineExceeded) {
+			fmt.Fprintf(a.LogWriter, "error: timeout waiting for response: %v\n", err)
+		} else if errors.Is(err, io.EOF) {
+			fmt.Fprintf(a.LogWriter, "error: connection lost: %v\n", err)
+		} else if err != nil {
+			fmt.Fprintf(a.LogWriter, "error: decoding response: %v\n", err)
+		} else {
+			_, authorized, err = a.channel.ConfirmClose(*m.Close)
+			if err != nil {
+				fmt.Fprintf(a.LogWriter, "error: confirming close response: %v\n", err)
+			}
+		}
+		closeReady <- authorized
+	}()
+	authorized := false
+	waitTime := observationPeriodTime * 2
+	select {
+	case authorized = <-closeReady:
+	case <-time.After(waitTime):
+	}
+	if authorized {
+		fmt.Fprintf(a.LogWriter, "close authorized\n")
+		_, immediateCloseTx, err := a.channel.CloseTxs()
+		if err != nil {
+			fmt.Fprintf(a.LogWriter, "error: building immediate close tx: %v", err)
+		} else {
+			closeTx = immediateCloseTx
+		}
+	} else {
+		remainingWaitTime := waitTime - time.Since(timeStart)
+		fmt.Fprintf(a.LogWriter, "close not authorized, waiting %v\n", remainingWaitTime)
+		time.Sleep(remainingWaitTime)
+	}
+	// Submit close tx
+	err = a.Submitter.SubmitFeeBumpTx(closeTx)
+	if err != nil {
+		return fmt.Errorf("submitting close tx: %w", err)
+	}
+	return nil
+}
+
+func (a *Agent) loop() {
+	var err error
+	recv := json.NewDecoder(io.TeeReader(a.Conn(), a.LogWriter))
+	send := json.NewEncoder(io.MultiWriter(a.Conn(), a.LogWriter))
+	for {
+		m := message{}
+		err = recv.Decode(&m)
+		if err != nil {
```

Comment:
❓ is every log that comes up in the console from the other party? Reason I'm wondering is the logs are pretty hard to mentally parse in the console while going through the steps of open/close/payment.

Could we add a log here that says something like `fmt.Println("receiving message:")`? I believe in here is where it should go

Reply:
The logging needs redoing to be able to do this. Right now logs are sent to stderr in between the network connection and the json decoder and if we add the println here it will print out after the message rather than before it. I think we can defer improving this for now and iterate on it in another PR later.

Reply:
sgtm 👍